fab/train_with_buffer.py

```diff
@@ -0,0 +1,148 @@
+        self.flow_device = next(model.flow.parameters()).device
+        self.clip_ais_weights_frac = clip_ais_weights_frac
+
+
+    def run(self,
+            n_iterations: int,
+            batch_size: int,
+            eval_batch_size: Optional[int] = None,
+            n_eval: Optional[int] = None,
+            n_plot: Optional[int] = None,
+            n_checkpoints: Optional[int] = None,
+            save: bool = True) -> None:
+
+
+        if save:
+            pathlib.Path(self.plots_dir).mkdir(exist_ok=True)
+            pathlib.Path(self.checkpoints_dir).mkdir(exist_ok=True)
+        if n_checkpoints:
+            checkpoint_iter = list(np.linspace(0, n_iterations - 1, n_checkpoints, dtype="int"))
+        if n_eval is not None:
+            eval_iter = list(np.linspace(0, n_iterations - 1, n_eval, dtype="int"))
+            assert eval_batch_size is not None
+        if n_plot is not None:
+            plot_iter = list(np.linspace(0, n_iterations - 1, n_plot, dtype="int"))
+
+        pbar = tqdm(range(n_iterations))
+        for i in pbar:
+            self.optimizer.zero_grad()
+            # collect samples and log weights with AIS.
+            x_ais, log_w_ais = self.model.\
+                annealed_importance_sampler.sample_and_log_weights(batch_size)
+            x_ais = x_ais.detach()
+            log_w_ais = log_w_ais.detach()
+            if self.clip_ais_weights_frac is not None:
+                # optional clipping of log weights
+                k = max(2, int(self.clip_ais_weights_frac * log_w_ais.shape[0]))
+                max_log_w = torch.min(torch.topk(log_w_ais, k, dim=0).values)
+                log_w_ais = torch.clamp_max(log_w_ais, max_log_w)
+
+            # perform one update using the recently collected AIS samples and log weights
+            loss = self.model.fab_alpha_div_loss_inner(x_ais, log_w_ais)
+            if not torch.isnan(loss) and not torch.isinf(loss):
+                loss.backward()
+                grad_norm = torch.nn.utils.clip_grad_norm_(self.model.parameters(),
+                                                           self.max_gradient_norm)
+                self.optimizer.step()
+                if self.optim_schedular:
+                    self.optim_schedular.step()
+            else:
+                print("nan loss in non-replay step")
+
+            # we log info from the step of the recently generated ais points.
+            info = self.model.get_iter_info()
+            info.update(loss=loss.cpu().detach().item(),
+                        step=i,
+                        grad_norm=grad_norm.cpu().detach().item())
+            self.logger.write(info)
+
+            # We now take an additinal self.n_batches_buffer_sampling gradient steps using
+            # data from the replay buffer.
+            for (x, log_w) in self.buffer.sample_n_batches(
+                    batch_size=batch_size, n_batches=self.n_batches_buffer_sampling):
```

Comment:
@VincentStimper these lines (72 - 119) will need to be equivalently updated for the dipeptide problem

Reply:
We unfortunately can no longer have a simple loop where we repeatedly just call `loss()`